Set maximum price of inverse payout curve to Olivia's max attestation price

luckysori opened the pull request:
Fix https://github.com/itchysats/itchysats/issues/2571.

This is a second attempt after the simple https://github.com/itchysats/itchysats/pull/2637 broke every protocol under the sun and we had to revert it.

The diff is large because we have to introduce new versions of several protocols to be able to introduce this small breaking change into the inverse payout curve implementation. This can be a good thing though, as pointed out by @bonomat, as it gives us an opportunity to introduce other breaking changes for these protocols before the next release.

luckysori committed:
refactor(payouts): Remove generic Payouts constructor

We remove it in favour of the previously private `new_inverse` and
`new_quanto`.

It's a little bit more code, but it allows us to be flexible with the
arguments that each constructor can take.

luckysori committed:
feat(rollover): Introduce /itchysats/rollover/3.0.0

This version is introduced to use the payout curve configuration of
`InverseMaxPrice::OliviaMax` for inverse payout curves.

It is not yet used, because we only intend to support two versions of
the same protocol at the same time. So we will first have to drop
version 1.0.0 and deprecate version 2.0.0.

luckysori committed:
feat!(rollover): Drop support for v1.0.0 in favour of v3.0.0

As things stand:

- `/itchysats/rollover/1.0.0` is dropped.
- `/itchysats/rollover/2.0.0` is deprecated.
- `/itchysats/rollover/3.0.0` is newly supported.

luckysori committed:
chore(rollover): Rename rollover modules

We "fix" the module names:

- The module `current` corresponds to the deprecated protocol, so we
rename it to deprecated.

- The module `next` corresponds to the current protocol, so we rename
it to `current`.

We do this in a separate commit to make it easier to review.

luckysori committed:
chore(rollover): Remove rollover Version enum

We now use separate protocol actors to support different versions of
protocols, so it makes sense to remove the `rollover::Version` enum
from that perspective.

We also only need to support the latest variant in that enum (because
all other variants correspond to the legacy network stack), which
means we can remove the enum altogether.

luckysori committed:
chore: Get rid of some libp2p prefixes

No need to specify that a protocol actor works over libp2p, because
they all do!

luckysori committed:
chore: Rename xtra-libp2p-ping in Cargo.toml's

The `xtra_libp2p_` prefix is noisy.

luckysori committed:
chore: Rename xtra-libp2p-offer in Cargo.toml's

The `xtra_libp2p_` prefix is noisy.

luckysori committed:
feat(collab-settlement): Add version 2.0.0

Currently both versions are equivalent. We will introduce new changes
int the next commit to make it easier to review.

luckysori committed:
feat(collab-settlement): Fix payout curve bug for version 2.0.0

Version 2.0.0 now uses the payout curve configuration of
`InverseMaxPrice::OliviaMax` for inverse payout curves, fixing the bug
mentioned in the commit title.

luckysori committed:
feat(order): Add version 2.0.0

Currently both versions are equivalent. We will introduce new changes
int the next commit to make it easier to review.

luckysori committed:
feat(order): Fix payout curve bug for version 2.0.0

Version 2.0.0 now uses the payout curve configuration of
`InverseMaxPrice::OliviaMax` for inverse payout curves, fixing the bug
mentioned in the commit title.

bonomat commented on model/src/payout_curve.rs:
```diff
@@ -76,8 +76,9 @@ impl Payouts {
         (leverage_long, leverage_short): (Leverage, Leverage),
         n_payouts: usize,
         fee: CompleteFee,
+        inverse_max_price_config: InverseMaxPrice,
```
What do you think of not exposing this but creating a new function something like `new_inverse_deprecated` which defaults to `InverseMaxPrice::DoubleOfInitial`. That way we don't need to expose the constant outside of this module and the caller does not need to know about it.

luckysori replied:
Yep, that's a better idea! Let me give it a go.

luckysori replied:
I did it everywhere except for a few places in `model` itself. 

And I marked `InverseMaxPrice` and `new_inverse` as `pub(crate)` to ensure that they're not used elsewhere.

bonomat commented on xtra-libp2p-rollover/src/next/taker.rs:
```diff
@@ -0,0 +1,368 @@
+            },
+        );
+    }
+}
```
I did not check the logic in here. I guess this is all copy&paste?

luckysori replied:
It should be a straight copy of the original `current`, apart from using `MaxInversePrice::OliviaMax`.

It is a limitation of our approach to backwards compatibility (creating multiple copies with some differences): we can't easily tell what the differences are.

bonomat commented on daemon/src/collab_settlement/current.rs:
```diff
@@ -0,0 +1,5 @@
+pub mod protocol;
+pub mod taker;
+
+pub const PROTOCOL: &str = "/itchysats/collab-settlement/2.0.0";
```
Since you introduce 0.2.0 we could go for a breaking change here and implement: https://github.com/itchysats/itchysats/issues/2924 :)

luckysori replied:
Sure, we should prioritise that ticket before we release something with these changes.

bonomat commented on model/src/payout_curve.rs:
```diff
@@ -86,6 +87,13 @@ impl Payouts {
+            let n_payouts = payouts.len() - 1;
+            let short_liquidation = payouts.get_mut(n_payouts).expect("several payouts");
+            short_liquidation.range =
+                *short_liquidation.range.start()..=maia_core::interval::MAX_PRICE_DEC;
```
You did not have to adapt a test for this?

luckysori replied:
Well, I didn't modify the inverse payout curve implementation. That's where most tests are located.

The `proptest` below still works after these changes because it doesn't check against hard-coded values.

bonomat reviewed (commented):
Added some initial comments.  Looking good so far.

luckysori committed:
feat(payouts): Allow configuration of inverse max price

With the new configuration enum `InverseMaxPrice` we can optionally
fix a bug[1] with the short liquidation events of the inverse payout
curve.

The reason why we don't unconditionally fix the bug is to maintain
backwards compatibility for all network protocols that depend on this
API. In fact, this patch sets the value to
`InverseMaxPrice::DoubleOfInitial` in all cases for now. A future
patch that introduces new versions of different network protocols will
introduce the use of the other variant.

[1]: https://github.com/itchysats/itchysats/issues/2571.

luckysori committed:
chore(shared-bin): Remove unused dependency

bonomat reviewed (approved):
Good work!
